Checkout: Add basic wpcom-checkout package

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

This adds a new npm package called `wpcom-checkout` which can be used as a toolbox for checkout in calypso (and possibly other places in the future). The advantages of moving code to this package is that such code will have strict type checking, as opposed to within calypso where types are allowed but not checked. It also provides a mechanism for that code to be shared with other WordPress.com checkout implementations outside of calypso (such as within wp-admin).

This PR moves only some line item translation functions to the package. These are functions which bridge the gap between the `shopping-cart` package and the `composite-checkout` package. Other code can be moved at a later time.

#### Testing instructions

This should not cause any visible changes to checkout, so mostly we just need to verify that nothing is broken. Some notable things to test which have been modified:

- Verify that translations for the words "Total", "Subtotal", and "Coupon" in the cart are still translated correctly to other locales.
- Verify that a purchase made through the PayPal Express payment method still works correctly.
- Verify that a purchase made through the existing card payment method still works correctly.
- Verify that when you have sufficient credits to cover a full transaction, you still only see the "Free" payment method.

sirbrillig committed:
Add basic wpcom-checkout package

sirbrillig committed:
Use LineItem translation functions in checkout

sirbrillig committed:
Add wpcom-checkout to client config and tsconfig

sirbrillig committed:
Fix types in composite-checkout OrderReviewLineItems

sirbrillig committed:
Convert PaymentMethodStep to use new translate functions

sirbrillig committed:
Remove export of old translate functions

sirbrillig committed:
Update filterAppropriatePaymentMethods to get required data from cart

sirbrillig committed:
Update payPalProcessor to get couponCode from cart

sirbrillig committed:
Update call to filterAppropriatePaymentMethods with new arguments

sirbrillig committed:
Send responseCart to payment processor functions

sirbrillig committed:
Remove non-products from CheckoutProvider items in checkout

sirbrillig committed:
Update WPCOMCart to remove unused properties

sirbrillig committed:
Remove unused data from translateResponseCartToWPCOMCart

sirbrillig committed:
Remove unused emptyWPCOMCart

sirbrillig committed:
Remove unused WPCOMCartItem variants

sirbrillig committed:
Remove non-product type filter in PayPal processor function

There are only products in there now.

sirbrillig committed:
Remove non-product filter from all payment processor functions

There are no longer non-products in items.

sirbrillig committed:
Update uses of doesPurchaseHaveFullCredits to use wpcom-checkout

sirbrillig committed:
Remove old doesPurchaseHaveFullCredits

sirbrillig committed:
Remove tests for removed properties

nbloomf commented on client/my-sites/checkout/composite-checkout/test/translate-cart.js:
```diff
@@ -129,27 +121,6 @@ describe( 'translateResponseCartToWPCOMCart', function () {
 			} );
 		} );
 
-		describe( 'taxes', function () {
```
It looks like we're losing these tests, which makes me a little nervous. Is it possible/sensible to move them to the new package?

sirbrillig replied:
That's a good point. They were testing the moved functions implicitly, which is why I didn't notice. I'll create some tests in the package.

sirbrillig replied:
I did better and added tests to cover the specific cases of the functions.

sirbrillig committed:
Add tests to wpcom-checkout

nbloomf commented:
Looks good! Translations work as expected:

![image](https://user-images.githubusercontent.com/9310939/111563429-f6620e00-8765-11eb-9b52-35dac64381d5.png)

I tested paypal, stored card, and full credits purchases with no issues.

sirbrillig committed:
Also clamp credits value and not just display value

This is a change from the current version of the function but I think
that it is a good idea and was _probably_ an oversight before.
